lib/Fhp/Segment/VPP/VopHelper.php

```diff
@@ -0,0 +1,135 @@
+    public static function createHKVPPForPollingRequest(BPD $bpd, VopPollingInfo $pollingInfo): HKVPPv1
+    {
+        $hkvpp = static::createHKVPPForInitialRequest($bpd);
+        $hkvpp->aufsetzpunkt = $pollingInfo->getAufsetzpunkt();
+        $hkvpp->pollingId = $pollingInfo->getPollingId();
+        return $hkvpp;
+    }
+
+    /**
+     * @param Message $response The response we just received from the server.
+     * @param int $hkvppSegmentNumber The number of the HKVPP segment in the request we had sent.
+     * @return ?VopPollingInfo If the response indicates that the Verification of Payee is still ongoing, such that the
+     *     client should keep polling the server to (actively) wait until the result is available, this function returns
+     *     a corresponding polling info object. If no polling is required, it returns null.
+     */
+    public static function checkPollingRequired(Message $response, int $hkvppSegmentNumber): ?VopPollingInfo
+    {
+        // Note: We determine whether polling is required purely based on the presence of the primary polling token (
+        // the Aufsetzpunkt is mandatory, the polling ID is optional).
+        // The specification also contains the code "3093 Namensabgleich ist noch in Bearbeitung", which could also be
+        // used to indicate that polling is required. But the specification does not mandate its use, and we have not
+        // observed banks using it consistently, so we don't rely on it here.
+        $aufsetzpunkt = $response->findRueckmeldung(Rueckmeldungscode::AUFSETZPUNKT, $hkvppSegmentNumber);
+        if ($aufsetzpunkt === null) {
+            return null;
+        }
+        /** @var HIVPPv1 $hivpp */
+        $hivpp = $response->findSegment(HIVPPv1::class);
+        if ($hivpp->vopId !== null || $hivpp->paymentStatusReport !== null) {
+            // Implementation note: If this ever happens, it could be related to $artDerLieferungPaymentStatusReport.
+            throw new UnexpectedResponseException('Got response with Aufsetzpunkt AND vopId/paymentStatusReport.');
+        }
+        return new VopPollingInfo(
+            $aufsetzpunkt->rueckmeldungsparameter[0],
+            $hivpp?->pollingId,
+            $hivpp?->wartezeitVorNaechsterAbfrage,
+        );
+    }
+
+    /**
+     * @param Message $response The response we just received from the server.
+     * @param int $hkvppSegmentNumber The number of the HKVPP segment in the request we had sent.
+     * @return ?VopConfirmationRequestImpl If the response contains a confirmation request for the user, it is returned,
+     *     otherwise null (which may imply that the action was executed without requiring confirmation).
+     */
+    public static function checkVopConfirmationRequired(
+        Message $response,
+        int $hkvppSegmentNumber,
+    ): ?VopConfirmationRequestImpl {
+        $rueckmeldung = $response->findRueckmeldung(
+            Rueckmeldungscode::VOP_ERGEBNIS_NAMENSABGLEICH_PRUEFEN,
```

Comment:
Bei Namensgleichheit, kommt dieser Code nicht. Stattdessen `VOP_KEINE_NAMENSABWEICHUNG`. D.h. aktuell erkennt FinTs nicht, dass eine Bestätigung benötigt wird.
Für den Fall hatte ich leider keinen Testfall angelegt.

Reply:
Alles klar, danke! In der Tat weist die Spezifikation darauf hin, dass die Rückmeldungscodes 3091 (und 3076) explizit angeben würden, falls die VOP-Bestätigung (bzw. der 2FA-Check) übersprungen werden könnten. Aber das scheint hier nicht der Fall zu sein.

Ich habe das jetzt korrigiert und neue Commits gepusht. Für die anderen beiden Fälle werde ich mal einen abgewandelten Unit-Test plus eine Implementierung basteln, wobei das diesen PR nicht unbedingt aufhalten muss.

Das erklärt wohl auch teilweise, was mit dem Hinweis "Der Ablauf wird grundsätzlich nur durch die Rückmeldungscodes gesteuert" gemeint war.